hal/inc/watchdog_hal.h

```diff
@@ -1,45 +1,145 @@
-/**
- ******************************************************************************
- * @file    watchdog_hal.h
- * @author  Matthew McGowan
- * @version V1.0.0
- * @date    25-Sept-2014
- * @brief
- ******************************************************************************
-  Copyright (c) 2013-2015 Particle Industries, Inc.  All rights reserved.
-
-  This library is free software; you can redistribute it and/or
-  modify it under the terms of the GNU Lesser General Public
-  License as published by the Free Software Foundation, either
-  version 3 of the License, or (at your option) any later version.
-
-  This library is distributed in the hope that it will be useful,
-  but WITHOUT ANY WARRANTY; without even the implied warranty of
-  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-  Lesser General Public License for more details.
-
-  You should have received a copy of the GNU Lesser General Public
-  License along with this library; if not, see <http://www.gnu.org/licenses/>.
- ******************************************************************************
+/*
+ * Copyright (c) 2022 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
  */
 
 #ifndef WATCHDOG_HAL_H
 #define	WATCHDOG_HAL_H
 
 #include <stdint.h>
 #include <stdbool.h>
+#include "hal_platform.h"
+#include "system_tick_hal.h"
+
+#if HAL_PLATFORM_HW_WATCHDOG
+
+#define HAL_WATCHDOG_VERSION    1
+
+typedef void (*hal_watchdog_on_expired_callback_t)(void* context);
+
+typedef enum hal_watchdog_instance_t {
+    HAL_WATCHDOG_INSTANCE1 = 0,
+    HAL_WATCHDOG_INSTANCE2 = 1,
+} hal_watchdog_instance_t;
+
+typedef enum hal_watchdog_capability_t {
+    WATCHDOG_CAPS_NONE              = 0x00,
+    WATCHDOG_CAPS_SOFT_RESET        = 0x01,
+    WATCHDOG_CAPS_INT               = 0x02,
```

Comment:
Let's be a bit more verbose in capabilities: `HAL_WATCHDOG_CAPS_RESET`, `HAL_WATCHDOG_NOTIFY`, `HAL_WATCHDOG_NOTIFY_ONLY`.

Reply:
I didn't get why we need `HAL_WATCHDOG_NOTIFY_ONLY`. If the watchdog does only support notifying, we just need to set the `HAL_WATCHDOG_NOTIFY` capability.

Reply:
Guohui and I discussed this separately. The gist is:

> We need to differentiate in caps whether we can just get notified without reset or like e.g. on nrf52840 we get notification BUT still get reset anyway. These are separate things, if you just or RESET and NOTIFY it's difficult to understand whether the reset has to happen or not.